pkg/bitmap/bitmap.go

```diff
@@ -0,0 +1,212 @@
+	blockNum, mask := i/64, uint64(1)<<(i%64)
+	// if blockNum is out of range, extend b.bitBlock
+	for x, y := int(blockNum), len(b.bitBlock); x >= y; x-- {
+		b.bitBlock = append(b.bitBlock, 0)
+	}
+	oldBlock := b.bitBlock[blockNum]
+	newBlock := oldBlock | mask
+	if oldBlock != newBlock {
+		b.bitBlock[blockNum] = newBlock
+		b.length++
+	}
+}
+
+// Remove i from the Bitmap
+func (b *Bitmap) Remove(i int32) {
+	blockNum, mask := i/64, uint64(1)<<(i%64)
+	oldBlock := b.bitBlock[blockNum]
+	newBlock := oldBlock &^ mask
+	if oldBlock != newBlock {
+		b.bitBlock[blockNum] = newBlock
+		b.length--
+	}
+}
+
+// Clone the Bitmap
+func (b *Bitmap) Clone() *Bitmap {
+	ptr := Bitmap{b.length, make([]uint64, len(b.bitBlock))}
+	copy(ptr.bitBlock, b.bitBlock[:])
+	return &ptr
+}
+
+// ReplaceWith replace the content of b with that of newBitmap
+func (b *Bitmap) ReplaceWith(newBitmap *Bitmap) {
+	b.length = newBitmap.length
+	b.bitBlock = make([]uint64, len(newBitmap.bitBlock))
+	copy(b.bitBlock, newBitmap.bitBlock[:])
+}
+
+// countLengthForBlocks count all 1 bits within bitblock of begin and that of end
+func countLengthForBlocks(bitBlock []uint64, begin int, end int) uint64 {
+	length := uint64(0)
+	if begin > end {
+		return length
+	}
+	beginBlock := begin / 64
+	endBlock := end / 64
+	for i := beginBlock; i <= endBlock; i++ {
+		length += uint64(bits.OnesCount64(bitBlock[i]))
+	}
+	return length
+}
+
+// flipBitmapRange flip the bits within range (begin and end)
+func flipBitmapRange(bitBlock []uint64, begin int, end int) {
+	if begin > end {
+		return
+	}
+	if begin == end {
+		blockNum, mask := begin/64, uint64(1)<<(begin%64)
+		bitBlock[blockNum] = bitBlock[blockNum] ^ mask
+		return
+	}
+	beginBlock := begin / 64
+	endBlock := end / 64
+	if beginBlock == endBlock {
+		bitBlock[endBlock] ^= ((^uint64(0) << uint(begin%64)) & ((uint64(1) << (uint(end) % 64 + 1))-1))
+	} else {
+		bitBlock[beginBlock] ^= ^(^uint64(0) << uint(begin%64))
+		for i := beginBlock; i < endBlock; i++ {
+			bitBlock[i] = ^bitBlock[i]
+		}
+		bitBlock[endBlock] ^= ((uint64(1) << (uint(end) % 64 + 1))-1)
+	}
+}
+
+// removeBitmapRange remove the bits within range (begin and end)
+func removeBitmapRange(bitBlock []uint64, begin int, end int) {
+	if begin > end {
+		return
+	}
+	if begin == end {
+		blockNum, mask := begin/64, uint64(1)<<(begin%64)
+		bitBlock[blockNum] = bitBlock[blockNum] &^ mask
+		return
+	}
+	beginBlock := begin / 64
+	endBlock := end / 64
+	if beginBlock == endBlock {
+		bitBlock[beginBlock] &= (((uint64(1) << uint(begin%64)) - 1) | ^((uint64(1) << (uint(end) % 64 + 1)) - 1))
+	} else {
+		bitBlock[beginBlock] &= ((uint64(1) << uint(begin%64)) - 1)
+		for i := beginBlock + 1; i < endBlock; i++ {
+			bitBlock[i] &= ^bitBlock[i]
+		}
+		bitBlock[endBlock] &= ^((uint64(1) << (uint(end) % 64 + 1)) - 1)
+	}
+}
+
+// RemoveRange remove bits within range (begin and end) for the Bitmap
+func (b *Bitmap) RemoveRange(begin int, end int) {
```

Comment:
nit: all `begin int, end int` arguments in all functions can be written as `begin, end int`

Reply:
Please also mention if this range is inclusive or exclusive. It looks like the range parameters are inclusive.

Reply:
> Please also mention if this range is inclusive or exclusive. It looks like the range parameters are inclusive.

Good suggestion.